### PR TITLE
Make TransformStreamDefaultController more robust

### PR DESCRIPTION
**Pull request by jasnell**

### CODE REVIEW

File: src/workerd/api/streams/standard.c++ [hunk 1]
```diff
@@ -3217,13 +3230,16 @@ jsg::Promise<void> TransformStreamDefaultController::close(jsg::Lock& js) {
...
-        // Allows for a graceful close of the readable side. Close will
-        // complete once all of the queued data is read or the stream
-        // errors.
-        readableController.close(js);
+          JSG_VISITABLE_LAMBDA((this,ref=JSG_THIS),(ref),(jsg::Lock& js) -> jsg::Promise<void> {
+        KJ_IF_MAYBE(readableController, tryGetReadableController()) {
+          // Allows for a graceful close of the readable side. Close will
+          // complete once all of the queued data is read or the stream
+          // errors.
+          readableController->close(js);
```
Comment (jclee) [inline src/workerd/api/streams/standard.c++:3238]: It looks like `tryGetReadableController()`/`tryGetWritableController()` return non-owning references...  I think that ensures that there is at least one thing holding an owning reference at the time of the `KJ_IF_MAYBE()`, but I'm not familiar enough with `.close()` or the other protected operations to know if this is sufficient to ensure that the remaining references are not freed during the execution of `.close()`.  If not, maybe these scopes should hold onto a temporary reference to the controller, too?

**Review by jclee — APPROVED** [review]

**Review by harrishancock — APPROVED** [review]

File: src/workerd/api/streams/standard.c++ (hunk 1 above)
Comment (jasnell) [reply src/workerd/api/streams/standard.c++:3238]: These are protected by the introduction of the WeakRef. If those go away then these will always safely return nullptr. These should be safe.

File: src/workerd/api/streams/standard.c++ [hunk 2]
```diff
@@ -3161,47 +3170,51 @@ void TransformStreamDefaultController::error(jsg::Lock& js, v8::Local<v8::Value>
...
+      auto chunkRef = js.v8Ref(chunk);
+      return KJ_ASSERT_NONNULL(maybeBackpressureChange).promise.whenResolved().then(js,
+          JSG_VISITABLE_LAMBDA((chunkRef = kj::mv(chunkRef), ref=JSG_THIS),
+                              (chunkRef, ref), (jsg::Lock& js) mutable -> jsg::Promise<void> {
+        KJ_IF_MAYBE(writableController, ref->tryGetWritableController()) {
+          KJ_IF_MAYBE(error, writableController->isErroring(js)) {
+            return js.rejectedPromise<void>(*error);
+          }
         }
-        KJ_ASSERT(writableController->isWritable());
```
[the remark this answers is not in the document]
Comment (jasnell) [reply src/workerd/api/streams/standard.c++]: It wasn't actually useful here. It was a left over from older development

File: src/workerd/api/streams/standard.h [hunk 3]
```diff
@@ -661,6 +661,20 @@ class WritableImpl {
 
 }  // namespace jscontroller
 
+template <typename T>
+class WeakRef: public kj::Refcounted {
```
[the remark this answers is not in the document]
Comment (jasnell) [reply src/workerd/api/streams/standard.h:665]: Yeah I plan to reconcile these soon but that'll be a separate bit of work

**Pull request merged**